src/Biogeochemistry.jl

```diff
@@ -0,0 +1,197 @@
+abstract type AbstractContinuousFormBiogeochemistry <: AbstractBiogeochemistry end
+
+@inline extract_biogeochemical_fields(i, j, k, grid, fields, names::NTuple{1}) =
+    @inbounds (fields[names[1]][i, j, k],)
+
+@inline extract_biogeochemical_fields(i, j, k, grid, fields, names::NTuple{2}) =
+    @inbounds (fields[names[1]][i, j, k],
+               fields[names[2]][i, j, k])
+
+@inline extract_biogeochemical_fields(i, j, k, grid, fields, names::NTuple{N}) where N =
+    @inbounds ntuple(n -> fields[names[n]][i, j, k], Val(N))
+
+@inline function biogeochemical_transition(i, j, k, grid, bgc::AbstractContinuousFormBiogeochemistry,
+                                           val_tracer_name, clock, fields)
+
+    names_to_extract = tuple(required_biogeochemical_tracers(bgc)...,
+                             required_biogeochemical_auxiliary_fields(bgc)...)
+
+    fields_ijk = extract_biogeochemical_fields(i, j, k, grid, fields, names_to_extract)
+
+    x = xnode(Center(), Center(), Center(), i, j, k, grid)
+    y = ynode(Center(), Center(), Center(), i, j, k, grid)
+    z = znode(Center(), Center(), Center(), i, j, k, grid)
+
+    return bgc(val_tracer_name, x, y, z, clock.time, fields_ijk...)
+end
+
+@inline (bgc::AbstractContinuousFormBiogeochemistry)(val_tracer_name, x, y, z, t, fields...) = zero(x)
+
+struct NoBiogeochemistry <: AbstractBiogeochemistry end
+
+tracernames(tracers) = keys(tracers)
+tracernames(tracers::Tuple) = tracers
+
+@inline function has_biogeochemical_tracers(fields::NamedTuple, required_fields, grid)
+
+    for field_name in required_fields
+        if !(field_name in keys(fields))
+            fields = merge(fields, NamedTuple{(field_name, )}((CenterField(grid), )))
+        end
+    end
+
+    return fields
+end
+
+@inline has_biogeochemical_tracers(user_tracer_names::Tuple, required_names, grid) =
+    tuple(user_tracer_names..., required_names...)
+
+"""
+    validate_biogeochemistry(tracers, auxiliary_fields, bgc, grid, clock)
+
+Ensure that `tracers` contains biogeochemical tracers and `auxiliary_fields`
+contains biogeochemical auxiliary fields (e.g. PAR).
+"""
+@inline function validate_biogeochemistry(tracers, auxiliary_fields, bgc, grid, clock)
+    req_tracers = required_biogeochemical_tracers(bgc)
+    tracers = has_biogeochemical_tracers(tracers, req_tracers, grid)
+    req_auxiliary_fields = required_biogeochemical_auxiliary_fields(bgc)
+
+    all(field ∈ tracernames(auxiliary_fields) for field in req_auxiliary_fields) ||
+        error("$(req_auxiliary_fields) must be among the list of auxiliary fields to use $(typeof(bgc).name.wrapper)")
+
+    # Return tracers and aux fields so that users may overload and
+    # define their own special auxiliary fields (e.g. PAR in test)
+    return tracers, auxiliary_fields 
+end
+
+required_biogeochemical_tracers(::NoBiogeochemistry) = ()
+required_biogeochemical_auxiliary_fields(bgc::AbstractBiogeochemistry) = ()
+
+#=
```

Comment:
I commented this out. Should we get rid of BasicBiogeochemistry?

Reply:
I don't think its that valuble to have and maintain

Reply:
ok!